### PR TITLE
[gen_l10n] Locale resolution integration test

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

Adds an integration test to ensure that when a locale with a country code is specified, those messages are used. When the messages are not specified in the arb file for that country, but exist for the general language, it resolves to use the general language's messages if they exist. 

I noticed that this doesn't work with plural messages in the test (it won't work in any existing app for that matter). For example, if there is a plural message in `app_en.arb` that is not available in `app_en_CA.arb`, it will cause the app to crash. It fails on: 
```

════════ Exception caught by widgets library ═══════════════════════════════════
The following ArgumentError was thrown building Home(dirty, dependencies: [_LocalizationsScope-[GlobalKey#b34d2]]):
Invalid argument(s): The 'other' named argument must be provided

The relevant error-causing widget was
    Home 
package:locale_bug/main.dart:44
When the exception was thrown, this was the stack...
```

This is likely occurring because app contains a plural message for the general language, but not for the specific country. This is likely because the `intl` library attempts a lookup on the plural message for the country's locale, but since it doesn't exist, the code in the `intl` library for `pluralLogic` fails and throws an exception. I'm not entirely sure what to do about this now, so I left those messages out of the test. cc/ @HansMuller 

## Related Issues

n/a

## Tests

I added the following tests:

Adds more test coverage.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 1]
```diff
@@ -35,7 +35,8 @@ dependencies:
 ''';
 
   @override
-  final String main = r'''
+  String get main => _main;
```
Comment (HansMuller) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: Why bother with this?

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 2]
```diff
@@ -246,4 +247,88 @@ void main() {
...
+      ],
+      home: Home(),
+    ),
+  );
+}
+''';
+
+  final String appEnCa = r'''
+{
+  "@@locale": "en_CA",
```
Comment (HansMuller) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: For now, en_CA.arb will need to contain all of the same messages as en.arb or just the plurals?

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart [hunk 3]
```diff
@@ -246,4 +247,88 @@ void main() {
...
+
+import 'l10n/app_localizations.dart';
+
+class Home extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    try {
+      print('get localizations');
+      final AppLocalizations localizations = AppLocalizations.of(context);
+      print('localizations: $localizations');
```
Comment (HansMuller) [inline packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: I don't think we want these print statements.

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart (hunk 2 above)
Comment (shihaohong) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: Just the plurals. I'll add them here with a comment, and we can remove 1-2 for testing once we have implemented our own message lookup

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart (hunk 3 above)
Comment (shihaohong) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: They were an accident, I left them in while debugging. Removed!

File: packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart (hunk 1 above)
Comment (shihaohong) [reply packages/flutter_tools/test/integration.shard/test_data/gen_l10n_project.dart]: I was trying to figure out how to get around the @override lint for fields. I'll update this to not be convoluted like this.

**Comment by shihaohong** [comment]
Closing in favor of https://github.com/flutter/flutter/pull/51375

**Pull request closed without merging**